modules/core/src/main/scala/org/tessellation/domain/snapshot/programs/Download.scala

```diff
@@ -49,28 +53,52 @@ object Download {
     globalSnapshotContextFns: GlobalSnapshotContextFunctions[F],
     nodeStorage: NodeStorage[F],
     consensus: SnapshotConsensus[F, GlobalIncrementalSnapshot, GlobalSnapshotContext, _],
-    peerSelect: PeerSelect[F]
+    peerSelect: PeerSelect[F],
+    config: AppConfig
   ): Download[F] = new Download[F] {
 
     val logger = Slf4jLogger.getLogger[F]
 
     val minBatchSizeToStartObserving: Long = 1L
     val observationOffset = NonNegLong(4L)
     val fetchSnapshotDelayBetweenTrials = 10.seconds
+    val checkJoinablePeersDelayBetweenAttempts = 1.minute
+    val retryCheckJoinablePeersMax = 10.minutes
+    val lowerSelectablePeersCountThreshold: PosInt = 2
+    val upperSelectablePeersCountThreshold: PosInt = 99
+    val fallbackSelectablePeersCountThreshold: PosInt = 1
 
     type DownloadResult = (Signed[GlobalIncrementalSnapshot], GlobalSnapshotInfo)
     type ObservationLimit = SnapshotOrdinal
 
-    def download: F[Unit] =
-      nodeStorage
-        .tryModifyState(NodeState.WaitingForDownload, NodeState.DownloadInProgress, NodeState.WaitingForObserving)(start)
-        .flatMap(observe)
-        .flatMap { result =>
-          val ((snapshot, context), observationLimit) = result
+    def waitForPeers(minPeers: PosInt): F[Set[Peer]] = retryingOnFailures[Set[Peer]](
+      limitRetriesByCumulativeDelay(
+        threshold = retryCheckJoinablePeersMax,
+        policy = constantDelay(checkJoinablePeersDelayBetweenAttempts)
+      ),
+      a => (a.size >= minPeers).pure[F],
+      (a, details) => logger.info(s"Discovered ${a.size} selectable peers, waiting for more: $details")
+    )(clusterStorage.getResponsivePeers.map(_.filter(_.state === Ready)))
```

Comment:
IMO waiting for peers shouldn't be executed inside the download but before the download.

Reply:
The delay code has been moved to a separate object, `DelayForPeers`. That is injected into `DownloadDaemon` and the delay is called there.